// @tuyau/client@0.2.10/build/index.js
// src/client.ts
import ky from "ky";

// src/request.ts
import { serialize } from "object-to-formdata";

// src/errors.ts
var TuyauHTTPError = class extends Error {
  constructor(status, value, response, request, cause) {
    const reason = status ? `status code ${status}` : "an unknown error";
    const message = response ? `Request failed with ${reason}: ${request?.method.toUpperCase()} /${request?.url}` : `Request failed with ${reason}`;
    super(message, { cause });
    this.status = status;
    this.value = value;
    this.name = "TuyauHTTPError";
    this.value = value;
  }
};
var TuyauNetworkError = class extends Error {
  constructor(cause, request) {
    const message = request ? `Network error: ${request.method.toUpperCase()} ${request.url}` : "Network error occurred";
    super(message, { cause });
    this.cause = cause;
    this.name = "TuyauNetworkError";
  }
};

// src/utils.ts
function buildSearchParams(query) {
  if (!query) return "";
  let stringified = "";
  const append = (key, value, isArray = false) => {
    if (value === void 0 || value === null) return;
    const encodedKey = encodeURIComponent(key);
    const encodedValue = encodeURIComponent(value);
    const keyValuePair = `${encodedKey}${isArray ? "[]" : ""}=${encodedValue}`;
    stringified += (stringified ? "&" : "?") + keyValuePair;
  };
  for (const [key, value] of Object.entries(query)) {
    if (!value) continue;
    if (Array.isArray(value)) {
      for (const v of value) {
        append(key, v, true);
      }
    } else {
      append(key, `${value}`);
    }
  }
  return stringified;
}
function camelCase(str) {
  return str.replace(/_([a-z])/g, (g) => g[1].toUpperCase());
}
function removeSlash(value) {
  return value.replace(/^\//, "");
}
function isObject(value) {
  return typeof value === "object" && !Array.isArray(value) && value !== null;
}

// src/request.ts
var isServer = typeof FileList === "undefined";
var isReactNative = typeof navigator !== "undefined" && navigator.product === "ReactNative";
var TuyauRequest = class {
  #options;
  constructor(options) {
    this.#options = options;
  }
  /**
   * Check if the value is a file
   */
  #isFile(v) {
    if (isReactNative && isObject(v) && v.uri) return true;
    if (isServer) return v instanceof Blob;
    return v instanceof FileList || v instanceof File;
  }
  /**
   * Check if the object has a file inside it
   */
  #hasFile(obj) {
    if (!obj) return false;
    return Object.values(obj).some((val) => {
      if (Array.isArray(val)) return val.some(this.#isFile);
      return this.#isFile(val);
    });
  }
  async #makeRequest() {
    try {
      let key = "json";
      let body = this.#options.body;
      if (!(body instanceof FormData) && this.#hasFile(body)) {
        body = serialize(body, { indices: true });
        key = "body";
      } else if (body instanceof FormData) {
        key = "body";
      }
      const isGetOrHead = ["get", "head"].includes(this.#options.method);
      const response = await this.#options.client[this.#options.method](
        removeSlash(this.#options.path),
        {
          searchParams: buildSearchParams(this.#options.queryOptions?.query || {}),
          [key]: !isGetOrHead ? body : void 0,
          ...this.#options.queryOptions
        }
      );
      let data;
      let error;
      const responseType = response.headers.get("Content-Type")?.split(";")[0];
      if (responseType === "application/json") {
        data = await response.json();
      } else if (responseType === "application/octet-stream") {
        data = await response.arrayBuffer();
      } else {
        data = await response.text();
      }
      const status = response.status;
      if (!response.ok) {
        error = new TuyauHTTPError(response.status, data, response, {
          url: removeSlash(this.#options.path),
          method: this.#options.method
        });
        data = void 0;
      }
      return { data, error, response, status };
    } catch (originalError) {
      const error = new TuyauNetworkError(originalError, {
        url: removeSlash(this.#options.path),
        method: this.#options.method
      });
      return { data: void 0, error, response: new Response(), status: 0 };
    }
  }
  /**
   * Make the request and return the response
   */
  then(resolve, reject) {
    return this.#makeRequest().then(resolve, reject);
  }
  /**
   * Unwrap will returns `response.data` if the request is successful
   * and will throw an error if `response.error` is set.
   */
  async unwrap() {
    const result = await this.#makeRequest();
    if (result.error) throw result.error;
    return result.data;
  }
};

// src/matchit.ts
var SEP = "/";
var STYPE = 0;
var PTYPE = 1;
var ATYPE = 2;
var OTYPE = 3;
var SLASH = 47;
var COLON = 58;
var ASTER = 42;
var QMARK = 63;
function strip(str) {
  if (str === SEP) return str;
  str.charCodeAt(0) === SLASH && (str = str.substring(1));
  const len = str.length - 1;
  return str.charCodeAt(len) === SLASH ? str.substring(0, len) : str;
}
function split(str) {
  return (str = strip(str)) === SEP ? [SEP] : str.split(SEP);
}
function isMatch(arr, obj, idx) {
  idx = arr[idx];
  if (obj.val === idx && obj.type === STYPE) {
    return true;
  }
  if (idx === SEP) {
    return obj.type > PTYPE;
  }
  if (obj.type === STYPE) {
    return false;
  }
  if (idx === "") {
    return obj.end === "" && (obj.matcher ? obj.matcher.test(idx) : true);
  }
  if (!idx) {
    return obj.end === "";
  }
  return idx.endsWith(obj.end) && (obj.matcher ? obj.matcher.test(idx) : true);
}
function match(str, all) {
  const segs = split(str);
  const len = segs.length;
  let l;
  let i = 0;
  let tmp;
  const fn = isMatch.bind(isMatch, segs);
  for (; i < all.length; i++) {
    tmp = all[i];
    if (
      // eslint-disable-next-line no-cond-assign
      (l = tmp.length) === len || l < len && tmp[l - 1].type === ATYPE || l > len && tmp[l - 1].type === OTYPE
    ) {
      if (tmp.every(fn)) return tmp;
    }
  }
  return [];
}
function parse(str, matchers) {
  if (str === SEP) {
    return [{ old: str, type: STYPE, val: str, end: "" }];
  }
  if (typeof matchers !== "object") {
    matchers = {};
  }
  let c;
  let x;
  let t;
  let sfx;
  let val;
  let nxt = strip(str);
  let i = -1;
  let j = 0;
  let len = nxt.length;
  const out = [];
  while (++i < len) {
    c = nxt.charCodeAt(i);
    if (c === COLON) {
      j = i + 1;
      t = PTYPE;
      x = 0;
      sfx = "";
      while (i < len && nxt.charCodeAt(i) !== SLASH) {
        c = nxt.charCodeAt(i);
        if (c === QMARK) {
          x = i;
          t = OTYPE;
        } else if (c === 46 && sfx.length === 0) {
          sfx = nxt.substring(x = i);
        }
        i++;
      }
      val = nxt.substring(j, x || i);
      const matcher = matchers[val];
      out.push({
        old: str,
        type: t,
        val,
        end: sfx,
        matcher: matcher && matcher.match,
        cast: matcher && matcher.cast
      });
      nxt = nxt.substring(i);
      len -= i;
      i = 0;
      continue;
    } else if (c === ASTER) {
      out.push({ old: str, type: ATYPE, val: nxt.substring(i), end: "" });
      continue;
    } else {
      j = i;
      while (i < len && nxt.charCodeAt(i) !== SLASH) {
        ++i;
      }
      val = nxt.substring(j, i);
      out.push({ old: str, type: STYPE, val, end: "" });
      nxt = nxt.substring(i);
      len -= i;
      i = j = 0;
    }
  }
  return out;
}
function exec(str, arr) {
  let i = 0;
  let x;
  let y;
  const segs = split(str);
  const out = {};
  for (; i < arr.length; i++) {
    x = segs[i];
    y = arr[i];
    if (x === SEP) continue;
    if (y.val === "*") {
      out[y.val] = segs.slice(i);
      break;
    }
    if (x !== void 0 && y.type | OTYPE === 2) {
      const value = x.replace(y.end, "");
      out[y.val] = y.cast ? y.cast(value) : value;
    }
  }
  return out;
}

// src/route.ts
var RouteHelper = class {
  constructor(baseUrl, routes, client) {
    this.baseUrl = baseUrl;
    this.routes = routes;
    this.client = client;
    this.#parsedRoutes = routes?.map((route) => parse(route.path));
  }
  #parsedRoutes;
  /**
   * Get a route by its name and throw an error if it doesn't exist
   */
  #getRouteByName(name) {
    const route = this.routes.find((route2) => route2.name === name);
    if (!route) throw new Error(`Route ${name} not found`);
    return route;
  }
  #getRouteByPath(path, onlyGet = false) {
    const route = this.routes.find(
      (route2) => route2.path === path && (onlyGet ? route2.method.includes("GET") : true)
    );
    if (!route) throw new Error(`Route ${path} not found`);
    return route;
  }
  /**
   * Build an URL using the given route path and replace the parameters with
   * the given values
   */
  #buildUrl(path, params) {
    let objParams = {};
    let arrayParams = [];
    if (Array.isArray(params)) {
      arrayParams = params;
    } else {
      const entries = Object.entries(params || {}).map(([key, value]) => [key, value]);
      objParams = Object.fromEntries(entries);
    }
    let index = 0;
    return path.replace(/:([A-Z_a-z]+)/g, (_, key) => {
      const camelCaseKey = camelCase(key);
      if (objParams[camelCaseKey]) return objParams[camelCaseKey].toString();
      if (arrayParams[index]) return arrayParams[index++];
      throw new Error(`Route ${path} is missing the ${camelCaseKey} parameter`);
    });
  }
  /**
   * Get the current location URL
   */
  #location() {
    const {
      host = "",
      pathname = "",
      search = ""
    } = typeof window !== "undefined" ? window.location : {};
    return { host, pathname, search };
  }
  /**
   * Check if the given params match another params object
   */
  #matchParams(current, params) {
    if (!("params" in params) || !params.params) return true;
    return Object.entries(params.params).every(([key, value]) => {
      return value.toString() === current[key].toString();
    });
  }
  /**
   * Check if the route name matches the given wildcard pattern
   *
   * @example
   * #wildcardMatch('users.*', { name: 'users.index' }) // true
   */
  #wildcardMatch(name, route) {
    const regex = new RegExp(`^${name.replace("*", ".*")}$`);
    return regex.test(route.name);
  }
  /**
   * Check if the current route matches the given route name and params
   * If no route name is passed, just return the current route name
   */
  $current(routeName, params) {
    const { pathname, search } = this.#location();
    const segments = match(pathname, this.#parsedRoutes);
    if (!routeName) return this.#getRouteByPath(segments[0].old, true).name;
    const route = this.#getRouteByPath(segments[0].old);
    if (!this.#wildcardMatch(routeName, route)) return false;
    if (segments[0].old !== route.path) return false;
    if (!params) return true;
    const parsedParams = exec(pathname, segments);
    const paramsObject = Object.fromEntries(
      Object.entries(parsedParams).map(([key, value]) => [camelCase(key), value])
    );
    const paramsMatch = this.#matchParams(paramsObject, params);
    const queryMatch = params.query ? new URLSearchParams(search).toString() === new URLSearchParams(buildSearchParams(params.query)).toString() : true;
    return paramsMatch && queryMatch;
  }
  /**
   * Make a request to a route by its name
   */
  $route(name, params) {
    const route = this.#getRouteByName(name);
    const result = route.method.reduce((result2, method) => {
      result2[`$${method.toLowerCase()}`] = (body, options = {}) => {
        const path = this.#buildUrl(route.path, params);
        const isGetOrHead = ["get", "head"].includes(method.toLowerCase());
        return new TuyauRequest({
          body,
          path,
          method: method.toLowerCase(),
          client: this.client,
          queryOptions: isGetOrHead ? body : options
        });
      };
      return result2;
    }, {});
    result.method = route.method;
    result.name = route.name;
    result.path = route.path;
    result.params = route.params;
    return result;
  }
  /**
   * Generate an URL for the given route name
   *
   * @example
   * tuyau.$url('users.index', { query: { page: 1 } })
   * tuyau.$url('users.show', { params: { id: 1 } })
   * tuyau.$url('users.show', { params: ['1'] })
   */
  $url(name, options) {
    const route = this.#getRouteByName(name);
    const pathParams = options?.params || {};
    const path = this.#buildUrl(route.path, pathParams);
    const url = new URL(path, this.baseUrl);
    url.search = buildSearchParams(options?.query || {});
    return url.toString();
  }
  /**
   * Check if a route with the given name exists
   *
   * You can use `*` as a wildcard character
   *
   * @example
   * tuyau.$has('users.*')
   * tuyau.$has('users.index')
   * tuyau.$has('users.*.show')
   *
   */
  $has(name) {
    if (!name.includes("*")) return this.routes.some((route) => route.name === name);
    return this.routes.some((route) => this.#wildcardMatch(name, route));
  }
};

// src/client.ts
var methods = ["get", "post", "put", "delete", "patch", "head"];
var prefixedMethods = methods.map((method) => `$${method}`);
function createProxy(options) {
  const { client, baseUrl, config, paths = [] } = options;
  const routeHelper = new RouteHelper(baseUrl, config.api?.routes, client);
  return new Proxy(() => {
  }, {
    get(_, param) {
      return createProxy({ ...options, paths: param === "index" ? paths : [...paths, param] });
    },
    apply(_, __, [body, queryOptions]) {
      const lastPath = paths.at(-1);
      if (paths.length === 1 && ["$has", "$route", "$url", "$current"].includes(lastPath)) {
        const method2 = lastPath;
        return routeHelper[method2](body, queryOptions);
      }
      if (lastPath === "$url") {
        const searchParams = buildSearchParams(body?.query);
        const url = new URL(paths.join("/").replace("/$url", ""), baseUrl);
        return url.toString() + searchParams;
      }
      const isMethodCall = prefixedMethods.includes(lastPath);
      if (!isMethodCall && typeof body === "object") {
        return createProxy({ ...options, paths: [...paths, Object.values(body)[0]] });
      }
      const method = paths[paths.length - 1].slice(1);
      const isGetOrHead = ["get", "head"].includes(method);
      return new TuyauRequest({
        body,
        client,
        method,
        path: paths.slice(0, -1).join("/"),
        queryOptions: isGetOrHead ? body : queryOptions
      });
    }
  });
}
function appendCsrfToken(request) {
  const xCsrfToken = globalThis.document?.cookie.split("; ").find((row) => row.startsWith("XSRF-TOKEN="));
  if (!xCsrfToken) return;
  request.headers.set("X-XSRF-TOKEN", decodeURIComponent(xCsrfToken.split("=")[1]));
}
function createTuyau(options) {
  options.plugins?.forEach((plugin) => plugin({ options }));
  const baseUrl = options.baseUrl;
  const client = ky.create({
    prefixUrl: baseUrl,
    throwHttpErrors: false,
    ...options,
    hooks: {
      ...options.hooks,
      beforeRequest: [...options.hooks?.beforeRequest || [], appendCsrfToken]
    }
  });
  return createProxy({ client, baseUrl, config: options });
}
export {
  TuyauHTTPError,
  TuyauNetworkError,
  createTuyau
};
