// @adonisjs-community/modules@1.1.0/build/commands/make/controller.js
import {
  stubsRoot
} from "../../chunk-LZWHXLM6.js";
import {
  COMMAND_PREFIX,
  MODULE_FLAG,
  checkModule
} from "../../chunk-DLHOZQRX.js";
import {
  __decorateClass
} from "../../chunk-ADS4GRIL.js";

// commands/make/controller.ts
import stringHelpers from "@adonisjs/core/helpers/string";
import MakeController from "@adonisjs/core/commands/make/controller";
import { flags } from "@adonisjs/core/ace";
var MMakeController = class extends MakeController {
  async run() {
    if (!this.module) {
      return super.run();
    }
    if (!checkModule(this.app, this.module)) {
      this.kernel.exec(`${COMMAND_PREFIX}:module`, [this.module]);
    }
    const codemods = await this.createCodemods();
    await codemods.makeUsingStub(stubsRoot, this.stubPath, {
      flags: this.parsed.flags,
      actions: this.actions?.map((action) => stringHelpers.camelCase(action)),
      entity: this.app.generators.createEntity(this.name),
      singular: this.singular
    });
  }
};
__decorateClass([
  flags.string(MODULE_FLAG)
], MMakeController.prototype, "module", 2);
export {
  MMakeController as default
};
