// @adonisjs-community/modules@1.1.0/build/commands/make/factory.js
import {
  stubsRoot
} from "../../chunk-LZWHXLM6.js";
import {
  COMMAND_PREFIX,
  MODULE_FLAG,
  checkModule
} from "../../chunk-DLHOZQRX.js";
import {
  __decorateClass
} from "../../chunk-ADS4GRIL.js";

// commands/make/factory.ts
import { flags } from "@adonisjs/core/ace";
import MakeFactory from "@adonisjs/lucid/commands/make_factory";
var MMakeFactory = class extends MakeFactory {
  async run() {
    if (!this.module) {
      return super.run();
    }
    if (!checkModule(this.app, this.module)) {
      this.kernel.exec(`${COMMAND_PREFIX}:module`, [this.module]);
    }
    const codemods = await this.createCodemods();
    await codemods.makeUsingStub(stubsRoot, "make/factory/main.stub", {
      flags: this.parsed.flags,
      entity: this.app.generators.createEntity(this.model),
      model: this.app.generators.createEntity(this.model),
      pathAlias: `#${this.module}/models`
    });
  }
};
__decorateClass([
  flags.string(MODULE_FLAG)
], MMakeFactory.prototype, "module", 2);
export {
  MMakeFactory as default
};
