// @adonisjs-community/modules@1.1.0/build/commands/make/middleware.js
import {
  stubsRoot
} from "../../chunk-LZWHXLM6.js";
import {
  COMMAND_PREFIX,
  MODULE_FLAG,
  checkModule
} from "../../chunk-DLHOZQRX.js";
import {
  __decorateClass
} from "../../chunk-ADS4GRIL.js";

// commands/make/middleware.ts
import { basename, extname, relative } from "path";
import { flags } from "@adonisjs/core/ace";
import MakeMiddleware from "@adonisjs/core/commands/make/middleware";
import stringHelpers from "@adonisjs/core/helpers/string";
import { slash } from "@adonisjs/core/helpers";
var MMakeMiddleware = class extends MakeMiddleware {
  async run() {
    if (!this.module) {
      return super.run();
    }
    if (!checkModule(this.app, this.module)) {
      this.kernel.exec(`${COMMAND_PREFIX}:module`, [this.module]);
    }
    const stackChoices = ["server", "router", "named"];
    if (!this.stack) {
      this.stack = await this.prompt.choice(
        "Under which stack you want to register the middleware?",
        stackChoices
      );
    }
    if (!stackChoices.includes(this.stack)) {
      this.exitCode = 1;
      this.logger.error(
        `Invalid middleware stack "${this.stack}". Select from "${stackChoices.join(", ")}"`
      );
      return;
    }
    const codemods = await this.createCodemods();
    const { destination } = await codemods.makeUsingStub(stubsRoot, this.stubPath, {
      flags: this.parsed.flags,
      entity: this.app.generators.createEntity(this.name)
    });
    const middlewareRelativePath = slash(
      relative(this.app.middlewarePath(), destination).replace(extname(destination), "")
    );
    const name = stringHelpers.camelCase(
      basename(middlewareRelativePath).replace(/_middleware$/, "")
    );
    await codemods.registerMiddleware(this.stack, [
      {
        name,
        path: `#middleware/${middlewareRelativePath}`
      }
    ]);
  }
};
__decorateClass([
  flags.string(MODULE_FLAG)
], MMakeMiddleware.prototype, "module", 2);
export {
  MMakeMiddleware as default
};
