// @adonisjs-community/modules@1.1.0/build/commands/make/test.js
import {
  stubsRoot
} from "../../chunk-LZWHXLM6.js";
import {
  COMMAND_PREFIX,
  MODULE_FLAG,
  checkModule
} from "../../chunk-DLHOZQRX.js";
import {
  __decorateClass
} from "../../chunk-ADS4GRIL.js";

// commands/make/test.ts
import { flags } from "@adonisjs/core/ace";
import MakeTest from "@adonisjs/core/commands/make/test";
var MMakeTest = class extends MakeTest {
  /**
   * Returns the suite name for creating the test file
   */
  async #getSuite() {
    if (this.suite) {
      return this.suite;
    }
    const rcFileSuites = this.app.rcFile.tests.suites;
    if (rcFileSuites.length === 1) {
      return rcFileSuites[0].name;
    }
    return this.prompt.choice(
      "Select the suite for the test file",
      this.app.rcFile.tests.suites.map((suite) => {
        return suite.name;
      }),
      {
        validate(choice) {
          return choice ? true : "Please select a suite";
        }
      }
    );
  }
  /**
   * Returns the directory path for the selected suite.
   */
  async #getSuiteDirectory(directories) {
    if (directories.length === 1) {
      return directories[0];
    }
    return this.prompt.choice("Select directory for the test file", directories, {
      validate(choice) {
        return choice ? true : "Please select a directory";
      }
    });
  }
  /**
   * Find suite info from the rcFile file
   */
  #findSuite(suiteName) {
    return this.app.rcFile.tests.suites.find((suite) => {
      return suite.name === suiteName;
    });
  }
  /**
   * Executed by ace
   */
  async run() {
    if (!this.module) {
      return super.run();
    }
    if (!checkModule(this.app, this.module)) {
      this.kernel.exec(`${COMMAND_PREFIX}:module`, [this.module]);
    }
    const suite = this.#findSuite(await this.#getSuite());
    if (!suite) {
      this.logger.error(`The "${this.suite}" suite is not configured inside the "adonisrc.js" file`);
      this.exitCode = 1;
      return;
    }
    const codemods = await this.createCodemods();
    await codemods.makeUsingStub(stubsRoot, this.stubPath, {
      flags: this.parsed.flags,
      entity: this.app.generators.createEntity(this.name),
      suite: {
        directory: await this.#getSuiteDirectory(suite.directories)
      }
    });
  }
};
__decorateClass([
  flags.string(MODULE_FLAG)
], MMakeTest.prototype, "module", 2);
export {
  MMakeTest as default
};
