// @tuyau/inertia@0.0.15/build/react/index.js
// src/react/index.tsx
import React from "react";
import { Link as InertiaLink, router as InertiaRouter } from "@inertiajs/react";
var TuyauContext = React.createContext(null);
function TuyauProvider(props) {
  return /* @__PURE__ */ React.createElement(TuyauContext.Provider, { value: props.client }, props.children);
}
var useTuyau = () => React.useContext(TuyauContext);
function LinkInner(props, ref) {
  const tuyau = useTuyau();
  if (!tuyau) throw new Error("You must wrap your app in a TuyauProvider");
  const route = tuyau.$route(props.route, props.params);
  const url = tuyau.$url(props.route, { params: props.params });
  return /* @__PURE__ */ React.createElement(
    InertiaLink,
    {
      ...props,
      route: void 0,
      params: void 0,
      href: url,
      method: route.method[0],
      ref
    }
  );
}
var Link = React.forwardRef(LinkInner);
function useRouter() {
  const tuyau = useTuyau();
  if (!tuyau) throw new Error("You must wrap your app in a TuyauProvider");
  const router = {
    visit: (props, options) => {
      const route = tuyau.$route(props.route, props.params);
      const url = tuyau.$url(props.route, { params: props.params });
      return InertiaRouter.visit(url, {
        ...options,
        method: route.method[0]
      });
    }
  };
  return router;
}
export {
  Link,
  TuyauProvider,
  useRouter,
  useTuyau
};
