// @japa/file-system@2.3.2/build/index.js
// index.ts
import { tmpdir } from "node:os";
import { join as join2 } from "node:path";
import { fileURLToPath } from "node:url";
import { createId } from "@paralleldrive/cuid2";
import { TestContext, Test } from "@japa/runner/core";

// src/assert.ts
import slash from "slash";
import { Assert } from "@japa/assert";
Assert.macro("fileExists", async function(filePath, message) {
  this.incrementAssertionsCount();
  const hasFile = await this.fs.exists(filePath);
  this.evaluate(hasFile, "expected #{this} file to exist", {
    thisObject: filePath,
    expected: "",
    actual: "",
    showDiff: false,
    prefix: message,
    operator: "exists"
  });
});
Assert.macro("fileNotExists", async function(filePath, message) {
  this.incrementAssertionsCount();
  const hasFile = await this.fs.exists(filePath);
  this.evaluate(!hasFile, "expected #{this} file to not exist", {
    thisObject: filePath,
    expected: "",
    actual: "",
    showDiff: false,
    prefix: message,
    operator: "notExists"
  });
});
Assert.macro(
  "fileEquals",
  async function(filePath, contents, message) {
    this.incrementAssertionsCount();
    const hasFile = await this.fs.exists(filePath);
    if (!hasFile) {
      this.evaluate(hasFile, "expected #{this} file to exist", {
        thisObject: filePath,
        expected: "",
        actual: "",
        showDiff: false,
        prefix: message,
        operator: "exists"
      });
      return;
    }
    const onDiskContents = await this.fs.contents(filePath);
    this.evaluate(onDiskContents === contents, "expected #{this} file contents to equal #{exp}", {
      thisObject: filePath,
      expected: contents,
      actual: onDiskContents,
      prefix: message,
      operator: "strictEqual"
    });
  }
);
Assert.macro(
  "fileContains",
  async function(filePath, substring, message) {
    this.incrementAssertionsCount();
    const hasFile = await this.fs.exists(filePath);
    if (!hasFile) {
      this.evaluate(hasFile, "expected #{this} file to exist", {
        thisObject: filePath,
        expected: "",
        actual: "",
        showDiff: false,
        prefix: message,
        operator: "exists"
      });
      return;
    }
    const onDiskContents = await this.fs.contents(filePath);
    if (typeof substring === "string") {
      this.evaluate(
        onDiskContents.includes(substring),
        "expected #{this} file contents to contain #{exp}",
        {
          thisObject: filePath,
          expected: substring,
          actual: onDiskContents,
          prefix: message,
          operator: "containsSubset"
        }
      );
      return;
    }
    if (Array.isArray(substring)) {
      substring.forEach((one) => {
        this.evaluate(
          onDiskContents.includes(one),
          "expected #{this} file contents to contain #{exp}",
          {
            thisObject: filePath,
            expected: one,
            actual: onDiskContents,
            prefix: message,
            operator: "containsSubset"
          }
        );
      });
      return;
    }
    this.evaluate(
      substring.test(onDiskContents),
      "expected #{this} file contents to match #{exp}",
      {
        thisObject: filePath,
        expected: substring,
        actual: onDiskContents,
        prefix: message,
        showDiff: false,
        operator: "strictEqual"
      }
    );
  }
);
Assert.macro(
  "fileNotContains",
  async function(filePath, substring, message) {
    this.incrementAssertionsCount();
    const hasFile = await this.fs.exists(filePath);
    if (!hasFile) {
      this.evaluate(hasFile, "expected #{this} file to exist", {
        thisObject: filePath,
        expected: "",
        actual: "",
        showDiff: false,
        prefix: message,
        operator: "exists"
      });
      return;
    }
    const onDiskContents = await this.fs.contents(filePath);
    if (typeof substring === "string") {
      this.evaluate(
        !onDiskContents.includes(substring),
        `expected #{this} file contents to not contain '${substring}'`,
        {
          thisObject: filePath,
          expected: substring,
          actual: onDiskContents,
          prefix: message,
          showDiff: false,
          operator: "containsSubset"
        }
      );
      return;
    }
    if (Array.isArray(substring)) {
      substring.forEach((one) => {
        this.evaluate(
          !onDiskContents.includes(one),
          `expected #{this} file contents to not contain '${one}'`,
          {
            thisObject: filePath,
            expected: one,
            actual: onDiskContents,
            prefix: message,
            showDiff: false,
            operator: "containsSubset"
          }
        );
      });
      return;
    }
  }
);
Assert.macro(
  "fileSameAs",
  async function(filePath, otherFilePath, message) {
    this.incrementAssertionsCount();
    const hasFile = await this.fs.exists(filePath);
    if (!hasFile) {
      this.evaluate(hasFile, "expected #{this} file to exist", {
        thisObject: filePath,
        expected: "",
        actual: "",
        showDiff: false,
        prefix: message,
        operator: "exists"
      });
      return;
    }
    const hasOtherFile = await this.fs.exists(otherFilePath);
    if (!hasOtherFile) {
      this.evaluate(hasOtherFile, "expected comparing file #{this} to exist", {
        thisObject: otherFilePath,
        expected: "",
        actual: "",
        showDiff: false,
        prefix: message,
        operator: "exists"
      });
      return;
    }
    const onDiskContents = await this.fs.contents(filePath);
    const otherFileContents = await this.fs.contents(otherFilePath);
    this.evaluate(
      onDiskContents === otherFileContents,
      "expected #{this} file contents to equal #{exp}",
      {
        thisObject: filePath,
        expected: otherFileContents,
        actual: onDiskContents,
        prefix: message,
        operator: "strictEqual"
      }
    );
  }
);
Assert.macro("fileIsEmpty", async function(filePath, message) {
  this.incrementAssertionsCount();
  const hasFile = await this.fs.exists(filePath);
  if (!hasFile) {
    this.evaluate(hasFile, "expected #{this} file to exist", {
      thisObject: filePath,
      expected: "",
      actual: "",
      showDiff: false,
      prefix: message,
      operator: "exists"
    });
    return;
  }
  const onDiskContents = await this.fs.contents(filePath);
  this.evaluate(onDiskContents.trim() === "", "expected #{this} file to be empty", {
    thisObject: filePath,
    expected: "",
    actual: onDiskContents,
    prefix: message,
    operator: "strictEqual"
  });
});
Assert.macro("fileIsNotEmpty", async function(filePath, message) {
  this.incrementAssertionsCount();
  const hasFile = await this.fs.exists(filePath);
  if (!hasFile) {
    this.evaluate(hasFile, "expected #{this} file to exist", {
      thisObject: filePath,
      expected: "",
      actual: "",
      showDiff: false,
      prefix: message,
      operator: "exists"
    });
    return;
  }
  const onDiskContents = await this.fs.contents(filePath);
  this.evaluate(onDiskContents.trim() !== "", "expected #{this} file to be not empty", {
    thisObject: filePath,
    expected: "",
    showDiff: false,
    actual: onDiskContents,
    prefix: message,
    operator: "notStrictEqual"
  });
});
Assert.macro("dirExists", async function(dirPath, message) {
  this.incrementAssertionsCount();
  const hasDir = await this.fs.exists(dirPath);
  this.evaluate(hasDir, "expected #{this} directory to exist", {
    thisObject: dirPath,
    expected: "",
    actual: "",
    showDiff: false,
    prefix: message,
    operator: "exists"
  });
});
Assert.macro("dirNotExists", async function(dirPath, message) {
  this.incrementAssertionsCount();
  const hasDir = await this.fs.exists(dirPath);
  this.evaluate(!hasDir, "expected #{this} directory to not exist", {
    thisObject: dirPath,
    expected: "",
    actual: "",
    showDiff: false,
    prefix: message,
    operator: "notExists"
  });
});
Assert.macro("hasFiles", async function(files, message) {
  this.incrementAssertionsCount();
  const directoryFiles = await this.fs.readDir();
  const directoryFilesPaths = directoryFiles.map((file) => slash(file.path));
  const hasAllFiles = files.every(
    (file) => directoryFilesPaths.find((directoryFilePath) => directoryFilePath === file)
  );
  this.evaluate(hasAllFiles, "expected file system to have #{exp} files", {
    thisObject: "",
    expected: files,
    actual: directoryFilesPaths,
    showDiff: true,
    prefix: message,
    operator: "deepStrictEqual"
  });
});
Assert.macro("doesNotHaveFiles", async function(files, message) {
  this.incrementAssertionsCount();
  const directoryFiles = await this.fs.readDir();
  const directoryFilesPaths = directoryFiles.map((file) => slash(file.path));
  const doesNotHaveAllFiles = files.every(
    (file) => !directoryFilesPaths.find((directoryFilePath) => directoryFilePath === file)
  );
  this.evaluate(doesNotHaveAllFiles, "expected file system to not have #{exp} files", {
    thisObject: "",
    expected: directoryFilesPaths,
    actual: files,
    showDiff: true,
    prefix: message,
    operator: "deepStrictNotEqual"
  });
});
Assert.macro("dirIsEmpty", async function(dirPath, message) {
  this.incrementAssertionsCount();
  const hasDir = dirPath ? await this.fs.exists(dirPath) : true;
  if (!hasDir) {
    this.evaluate(hasDir, "expected #{this} directory to exist", {
      thisObject: dirPath,
      expected: "",
      actual: "",
      showDiff: false,
      prefix: message,
      operator: "exists"
    });
    return;
  }
  const directoryFiles = await this.fs.readDir(dirPath);
  const isEmpty = directoryFiles.length === 0;
  if (!isEmpty && dirPath) {
    this.evaluate(isEmpty, "expected #{this} directory to be empty", {
      thisObject: dirPath,
      expected: [],
      actual: directoryFiles.map(({ path }) => slash(path)),
      showDiff: true,
      prefix: message,
      operator: "deepStrictEqual"
    });
    return;
  }
  if (!isEmpty) {
    this.evaluate(isEmpty, "expected file system base directory to be empty", {
      thisObject: "",
      expected: [],
      actual: directoryFiles.map(({ path }) => slash(path)),
      showDiff: true,
      prefix: message,
      operator: "deepStrictEqual"
    });
  }
});
Assert.macro("dirIsNotEmpty", async function(dirPath, message) {
  this.incrementAssertionsCount();
  const hasDir = dirPath ? await this.fs.exists(dirPath) : true;
  if (!hasDir) {
    this.evaluate(hasDir, "expected #{this} directory to exist", {
      thisObject: dirPath,
      expected: "",
      actual: "",
      showDiff: false,
      prefix: message,
      operator: "exists"
    });
    return;
  }
  const directoryFiles = await this.fs.readDir(dirPath);
  const hasFiles = directoryFiles.length > 0;
  if (!hasFiles && dirPath) {
    this.evaluate(hasFiles, "expected #{this} directory to be not empty", {
      thisObject: dirPath,
      expected: [],
      actual: [],
      showDiff: false,
      prefix: message,
      operator: "deepStrictNotEqual"
    });
    return;
  }
  if (!hasFiles) {
    this.evaluate(hasFiles, "expected file system base directory to be not empty", {
      thisObject: "",
      expected: [],
      actual: [],
      showDiff: false,
      prefix: message,
      operator: "deepStrictNotEqual"
    });
  }
});

// src/file_system.ts
import { pathToFileURL } from "node:url";
import { dirname, join } from "node:path";
import Macroable from "@poppinss/macroable";
import { readdirpPromise } from "readdirp";
import { access, mkdir, readFile, rm, stat, writeFile } from "node:fs/promises";
import {
  constants
} from "node:fs";
var FileSystem = class extends Macroable {
  /**
   * Base URL to the directory for reading and writing
   * files
   */
  baseUrl;
  /**
   * Base path to the directory for reading and writing
   * files
   */
  basePath;
  constructor(basePath) {
    super();
    this.basePath = basePath;
    this.baseUrl = pathToFileURL(`${this.basePath}/`);
  }
  #makePath(filePath) {
    return join(this.basePath, filePath);
  }
  /**
   * Cleanup directory
   */
  async cleanup(options) {
    return rm(this.basePath, { recursive: true, force: true, maxRetries: 10, ...options });
  }
  /**
   * Creates a directory inside the root of the filesystem
   * path. You may use this method to create nested
   * directories as well.
   */
  mkdir(dirPath, options) {
    return mkdir(this.#makePath(dirPath), { recursive: true, ...options });
  }
  /**
   * Create a new file
   */
  async create(filePath, contents, options) {
    const absolutePath = this.#makePath(filePath);
    await mkdir(dirname(absolutePath), { recursive: true });
    return writeFile(this.#makePath(filePath), contents, options);
  }
  /**
   * Remove a file
   */
  async remove(filePath, options) {
    return rm(this.#makePath(filePath), { recursive: true, force: true, maxRetries: 2, ...options });
  }
  /**
   * Check if the root of the filesystem exists
   */
  async rootExists() {
    try {
      await access(this.basePath, constants.F_OK);
      return true;
    } catch (error) {
      if (error.code === "ENOENT") {
        return false;
      }
      throw error;
    }
  }
  /**
   * Check if a file exists
   */
  async exists(filePath) {
    try {
      await access(this.#makePath(filePath), constants.F_OK);
      return true;
    } catch (error) {
      if (error.code === "ENOENT") {
        return false;
      }
      throw error;
    }
  }
  /**
   * Returns file contents
   */
  async contents(filePath) {
    return readFile(this.#makePath(filePath), "utf-8");
  }
  /**
   * Dumps file contents to the stdout
   */
  async dump(filePath) {
    console.log("------------------------------------------------------------");
    console.log(`file path => "${filePath}"`);
    console.log(`contents => "${await this.contents(filePath)}"`);
  }
  /**
   * Returns stats for a file
   */
  async stats(filePath, options) {
    return stat(this.#makePath(filePath), options);
  }
  /**
   * Recursively reads files from a given directory
   */
  readDir(dirPath) {
    return dirPath ? readdirpPromise(this.#makePath(dirPath), { type: "files" }) : readdirpPromise(this.basePath, { type: "files" });
  }
  /**
   * Create a json file
   */
  async createJson(filePath, contents, options) {
    if (options && typeof options === "object") {
      const { replacer, spaces, ...rest } = options;
      return this.create(filePath, JSON.stringify(contents, replacer, spaces), rest);
    }
    return this.create(filePath, JSON.stringify(contents), options);
  }
  /**
   * Read and parse a json file
   */
  async contentsJson(filePath) {
    const contents = await readFile(this.#makePath(filePath), "utf-8");
    return JSON.parse(contents);
  }
};

// index.ts
function fileSystem(options) {
  const normalizeOptions = Object.assign(
    { basePath: join2(tmpdir(), createId()), autoClean: true },
    options
  );
  const baseRoot = typeof normalizeOptions.basePath === "string" ? normalizeOptions.basePath : fileURLToPath(normalizeOptions.basePath);
  const fsPlugin = function() {
    TestContext.getter("fs", () => new FileSystem(baseRoot), true);
    Test.executing((test) => {
      test.context.assert.fs = test.context.fs;
      if (normalizeOptions.autoClean) {
        test.context.cleanup(() => test.context.fs.cleanup());
      }
    });
  };
  return fsPlugin;
}
export {
  FileSystem,
  fileSystem
};
