// @tuyau/inertia@0.0.15/build/vue/index.js
// src/vue/index.tsx
import { defineComponent, h, inject } from "vue";
import { Link as InertiaLink, router as InertiaRouter } from "@inertiajs/vue3";
var TUYAU_PLUGIN = "TUYAU_PLUGIN";
function getClientKey(key) {
  const suffix = key ? `:${key}` : "";
  return `${TUYAU_PLUGIN}${suffix}`;
}
var TuyauPlugin = {
  install: (app, options) => {
    app.provide(getClientKey(), options.client);
  }
};
var Link = defineComponent(
  (props, ctx) => {
    const tuyau = inject(getClientKey());
    if (!tuyau) throw new Error("You must install the TuyauPlugin before using Link");
    const route = tuyau.$route(props.route, props.params);
    const url = tuyau.$url(props.route, { params: props.params });
    return () => h(
      InertiaLink,
      {
        ...props,
        route: void 0,
        params: void 0,
        href: url,
        method: route.method[0]
      },
      ctx.slots
    );
  },
  { props: ["route", "params"] }
);
function useRouter() {
  const tuyau = inject(getClientKey());
  if (!tuyau) throw new Error("You must install the TuyauPlugin before using useRouter");
  const router = {
    visit: (props, options) => {
      const route = tuyau.$route(props.route, props.params);
      const url = tuyau.$url(props.route, { params: props.params });
      return InertiaRouter.visit(url, {
        ...options,
        method: route.method[0]
      });
    }
  };
  return router;
}
export {
  Link,
  TUYAU_PLUGIN,
  TuyauPlugin,
  getClientKey,
  useRouter
};
