// @verrou/core@0.5.1/build/chunk-MDY27ECV.js
import {
  E_LOCK_NOT_OWNED
} from "./chunk-QYHBKL3W.js";

// src/drivers/database.ts
var DatabaseStore = class {
  #adapter;
  #initialized;
  constructor(adapter, config) {
    this.#adapter = adapter;
    this.#adapter.setTableName(config.tableName || "verrou");
    if (config.autoCreateTable !== false) {
      this.#initialized = this.#adapter.createTableIfNotExists();
    } else {
      this.#initialized = Promise.resolve();
    }
  }
  /**
   * Compute the expiration date based on the provided TTL
   */
  #computeExpiresAt(ttl) {
    if (ttl) return Date.now() + ttl;
    return null;
  }
  /**
   * Get the current owner of given lock key
   */
  async #getCurrentOwner(key) {
    await this.#initialized;
    const lock = await this.#adapter.getLock(key);
    return lock?.owner;
  }
  /**
   * Save the lock in the store if not already locked by another owner
   *
   * We basically rely on primary key constraint to ensure the lock is
   * unique.
   *
   * If the lock already exists, we check if it's expired. If it is, we
   * update it with the new owner and expiration date.
   */
  async save(key, owner, ttl) {
    await this.#initialized;
    try {
      await this.#adapter.insertLock({ key, owner, expiration: this.#computeExpiresAt(ttl) });
      return true;
    } catch (error) {
      const updatedRows = await this.#adapter.acquireLock({
        key,
        owner,
        expiration: this.#computeExpiresAt(ttl)
      });
      return updatedRows > 0;
    }
  }
  /**
   * Delete the lock from the store if it is owned by the owner
   * Otherwise throws a E_LOCK_NOT_OWNED error
   */
  async delete(key, owner) {
    const currentOwner = await this.#getCurrentOwner(key);
    if (currentOwner !== owner) throw new E_LOCK_NOT_OWNED();
    await this.#adapter.deleteLock(key, owner);
  }
  /**
   * Force delete the lock from the store. No check is made on the owner
   */
  async forceDelete(key) {
    await this.#adapter.deleteLock(key);
  }
  /**
   * Extend the lock expiration. Throws an error if the lock is not owned by the owner
   * Duration is in milliseconds
   */
  async extend(key, owner, duration) {
    await this.#initialized;
    const updated = await this.#adapter.extendLock(key, owner, duration);
    if (updated === 0) throw new E_LOCK_NOT_OWNED();
  }
  /**
   * Check if the lock exists
   */
  async exists(key) {
    await this.#initialized;
    const lock = await this.#adapter.getLock(key);
    if (!lock) return false;
    if (lock.expiration === null) return true;
    return lock.expiration > Date.now();
  }
};

export {
  DatabaseStore
};
