// @verrou/core@0.5.1/build/chunk-AW5FCDQ4.js
// src/lock.ts
import { setTimeout } from "node:timers/promises";
import { InvalidArgumentsException } from "@poppinss/utils/exceptions";

// src/helpers.ts
import string from "@poppinss/utils/string";
function resolveDuration(duration, defaultValue = 3e4) {
  if (typeof duration === "number") return duration;
  if (duration === null) return void 0;
  if (duration === void 0) {
    if (typeof defaultValue === "number") return defaultValue;
    if (typeof defaultValue === "string") return string.milliseconds.parse(defaultValue);
    return void 0;
  }
  return string.milliseconds.parse(duration);
}

// src/lock.ts
var Lock = class {
  #key;
  #owner;
  #lockStore;
  #ttl = null;
  #config;
  #expirationTime = null;
  constructor(key, lockStore, config, owner, ttl, expirationTime) {
    this.#key = key;
    this.#config = config;
    this.#ttl = ttl ?? null;
    this.#lockStore = lockStore;
    this.#expirationTime = expirationTime ?? null;
    this.#owner = owner ?? this.#generateOwner();
  }
  /**
   * Generate a new owner ID
   */
  #generateOwner() {
    return Math.random().toString(36).slice(2);
  }
  /**
   * Returns the owner ID
   */
  getOwner() {
    return this.#owner;
  }
  /**
   * Serialize the lock
   */
  serialize() {
    return {
      key: this.#key,
      owner: this.#owner,
      ttl: this.#ttl,
      expirationTime: this.#expirationTime
    };
  }
  /**
   * Acquire the lock
   */
  async acquire(options = {}) {
    this.#expirationTime = null;
    let attemptsDone = 0;
    const attemptsMax = options.retry?.attempts ?? this.#config.retry.attempts;
    const delay = resolveDuration(options.retry?.delay, this.#config.retry.delay);
    const timeout = resolveDuration(options.retry?.timeout, this.#config.retry.timeout);
    const start = Date.now();
    while (attemptsDone++ < attemptsMax) {
      const now = Date.now();
      const result = await this.#lockStore.save(this.#key, this.#owner, this.#ttl);
      if (result) {
        this.#expirationTime = this.#ttl ? now + this.#ttl : null;
        break;
      }
      if (attemptsDone === attemptsMax) return false;
      const elapsed = Date.now() - start;
      if (timeout && elapsed > timeout) return false;
      await setTimeout(delay);
    }
    this.#config.logger.debug({ key: this.#key }, "Lock acquired");
    return true;
  }
  /**
   * Try to acquire the lock immediately or throw an error
   */
  async acquireImmediately() {
    const result = await this.#lockStore.save(this.#key, this.#owner, this.#ttl);
    if (!result) return false;
    this.#expirationTime = this.#ttl ? Date.now() + this.#ttl : null;
    this.#config.logger.debug({ key: this.#key }, "Lock acquired with acquireImmediately()");
    return true;
  }
  /**
   * Acquire the lock, run the callback and release the lock automatically
   * after the callback is done.
   * Also returns the callback return value
   */
  async run(callback) {
    const handle = await this.acquire();
    if (!handle) return [false, null];
    try {
      const result = await callback();
      return [true, result];
    } finally {
      await this.release();
    }
  }
  /**
   * Same as `run` but try to acquire the lock immediately
   * Or throw an error if the lock is already acquired
   */
  async runImmediately(callback) {
    const handle = await this.acquireImmediately();
    if (!handle) return [false, null];
    try {
      const result = await callback();
      return [true, result];
    } finally {
      if (handle) await this.release();
    }
  }
  /**
   * Force release the lock
   */
  async forceRelease() {
    await this.#lockStore.forceDelete(this.#key);
  }
  /**
   * Release the lock
   */
  async release() {
    await this.#lockStore.delete(this.#key, this.#owner);
    this.#config.logger.debug({ key: this.#key }, "Lock released");
  }
  /**
   * Returns true if the lock is expired
   */
  isExpired() {
    if (this.#expirationTime === null) return false;
    return this.#expirationTime < Date.now();
  }
  /**
   * Get the remaining time before the lock expires
   */
  getRemainingTime() {
    if (this.#expirationTime === null) return null;
    return this.#expirationTime - Date.now();
  }
  /**
   * Extends the lock TTL
   */
  async extend(ttl) {
    const resolvedTtl = ttl ? resolveDuration(ttl) : this.#ttl;
    if (!resolvedTtl) throw new InvalidArgumentsException("Cannot extend a lock without TTL");
    const now = Date.now();
    await this.#lockStore.extend(this.#key, this.#owner, resolvedTtl);
    this.#expirationTime = now + resolvedTtl;
    this.#config.logger.debug({ key: this.#key, newTtl: this.#expirationTime }, "Lock extended");
  }
  /**
   * Returns true if the lock is currently locked
   */
  async isLocked() {
    return await this.#lockStore.exists(this.#key);
  }
};

// src/verrou.ts
import { noopLogger as noopLogger2 } from "typescript-log";

// src/lock_factory.ts
import { noopLogger } from "typescript-log";
var LockFactory = class _LockFactory {
  /**
   * Default configuration values
   */
  static #kDefaults = {
    ttl: 3e4,
    retry: {
      attempts: Number.POSITIVE_INFINITY,
      delay: 250,
      timeout: void 0
    }
  };
  /**
   * The store used to persist locks
   */
  #store;
  /**
   * Resolved LockFactory configuration
   */
  #config;
  constructor(store, options = {}) {
    this.#store = store;
    this.#config = {
      retry: {
        attempts: options.retry?.attempts ?? _LockFactory.#kDefaults.retry.attempts,
        delay: resolveDuration(options.retry?.delay, null) ?? _LockFactory.#kDefaults.retry.delay,
        timeout: resolveDuration(options.retry?.timeout, null)
      },
      logger: (options.logger ?? noopLogger()).child({ pkg: "verrou" })
    };
  }
  /**
   * Create a new lock
   */
  createLock(name, ttl = _LockFactory.#kDefaults.ttl) {
    return new Lock(name, this.#store, this.#config, void 0, resolveDuration(ttl));
  }
  /**
   * Restore a lock from a previous owner. This is particularly useful
   * if you want to release a lock from a different process than the one
   * that acquired it.
   */
  restoreLock(lock) {
    return new Lock(lock.key, this.#store, this.#config, lock.owner, lock.ttl, lock.expirationTime);
  }
};

// src/verrou.ts
var Verrou = class {
  /**
   * The name of the default store
   */
  #defaultStoreName;
  /**
   * The list of stores
   */
  #stores;
  /**
   * A cache of store instances. We are re-using the same store instance
   * instead of creating a new one every time
   */
  #storesCache = /* @__PURE__ */ new Map();
  /**
   * Logger instance
   */
  #logger;
  constructor(config) {
    this.#stores = config.stores;
    this.#defaultStoreName = config.default;
    this.#logger = (config.logger ?? noopLogger2()).child({ pkg: "verrou" });
  }
  /**
   * Access a store by its name
   */
  use(store) {
    const storeToUse = store ?? this.#defaultStoreName;
    if (!storeToUse) throw new Error("No store provided");
    if (this.#storesCache.has(storeToUse)) {
      return this.#storesCache.get(storeToUse);
    }
    const factory = new LockFactory(this.#stores[storeToUse].driver.factory(), {
      logger: this.#logger
    });
    this.#storesCache.set(storeToUse, factory);
    return factory;
  }
  /**
   * Create a new lock using the default store
   */
  createLock(name, ttl) {
    return this.use(this.#defaultStoreName).createLock(name, ttl);
  }
  /**
   * Restore a lock from a previous owner. This is particularly useful
   * if you want to release a lock from a different process than the one
   * that acquired it.
   */
  restoreLock(lock) {
    return this.use(this.#defaultStoreName).restoreLock(lock);
  }
};

export {
  Lock,
  LockFactory,
  Verrou
};
