// @verrou/core@0.5.1/build/src/drivers/dynamodb.js
import {
  E_LOCK_NOT_OWNED
} from "../../chunk-QYHBKL3W.js";

// src/drivers/dynamodb.ts
import {
  ConditionalCheckFailedException,
  CreateTableCommand,
  DeleteItemCommand,
  GetItemCommand,
  PutItemCommand,
  ResourceInUseException
} from "@aws-sdk/client-dynamodb";
function dynamodbStore(config) {
  return { factory: () => new DynamoDBStore(config) };
}
var DynamoDBStore = class {
  #initialized;
  /**
   * DynamoDB connection
   */
  #connection;
  /**
   * DynamoDB table name
   */
  #tableName;
  constructor(config) {
    this.#tableName = config.table.name;
    this.#connection = config.connection;
    this.#initialized = this.#createTableIfNotExists();
  }
  /**
   * Automatically create the Dynamo table if it does not exist.
   */
  async #createTableIfNotExists() {
    const command = new CreateTableCommand({
      TableName: this.#tableName,
      KeySchema: [{ AttributeName: "key", KeyType: "HASH" }],
      AttributeDefinitions: [{ AttributeName: "key", AttributeType: "S" }],
      ProvisionedThroughput: {
        ReadCapacityUnits: 4,
        WriteCapacityUnits: 4
      }
    });
    try {
      await this.#connection.send(command);
    } catch (error) {
      if (error instanceof ResourceInUseException) return;
      throw error;
    }
  }
  /**
   * Save the lock in the store if not already locked by another owner
   */
  async save(key, owner, ttl) {
    await this.#initialized;
    try {
      const command = new PutItemCommand({
        TableName: this.#tableName,
        Item: {
          key: { S: key },
          owner: { S: owner },
          ...ttl ? { expires_at: { N: (Date.now() + ttl).toString() } } : {}
        },
        ConditionExpression: "attribute_not_exists(#key) OR #expires_at < :now",
        ExpressionAttributeNames: {
          "#key": "key",
          "#expires_at": "expires_at"
        },
        ExpressionAttributeValues: { ":now": { N: Date.now().toString() } }
      });
      const result = await this.#connection.send(command);
      return result.$metadata.httpStatusCode === 200;
    } catch (err) {
      if (err instanceof ConditionalCheckFailedException) return false;
      throw err;
    }
  }
  /**
   * Delete the lock from the store if it is owned by the owner
   * Otherwise throws a E_LOCK_NOT_OWNED error
   */
  async delete(key, owner) {
    const command = new DeleteItemCommand({
      TableName: this.#tableName,
      Key: { key: { S: key } },
      ConditionExpression: "#owner = :owner",
      ExpressionAttributeNames: { "#owner": "owner" },
      ExpressionAttributeValues: { ":owner": { S: owner } }
    });
    try {
      await this.#connection.send(command);
    } catch (err) {
      throw new E_LOCK_NOT_OWNED();
    }
  }
  /**
   * Force delete the lock from the store. No check is made on the owner
   */
  async forceDelete(key) {
    const command = new DeleteItemCommand({
      TableName: this.#tableName,
      Key: { key: { S: key } }
    });
    await this.#connection.send(command);
  }
  /**
   * Check if the lock exists
   */
  async exists(key) {
    await this.#initialized;
    const command = new GetItemCommand({
      TableName: this.#tableName,
      Key: { key: { S: key } }
    });
    const result = await this.#connection.send(command);
    const isExpired = result.Item?.expires_at?.N && result.Item.expires_at.N < Date.now().toString();
    return result.Item !== void 0 && !isExpired;
  }
  /**
   * Extend the lock expiration. Throws an error if the lock is not owned by the owner
   * Duration is in milliseconds
   */
  async extend(key, owner, duration) {
    const command = new PutItemCommand({
      TableName: this.#tableName,
      Item: {
        key: { S: key },
        owner: { S: owner },
        expires_at: { N: (Date.now() + duration).toString() }
      },
      ConditionExpression: "#owner = :owner",
      ExpressionAttributeNames: { "#owner": "owner" },
      ExpressionAttributeValues: { ":owner": { S: owner } }
    });
    try {
      await this.#connection.send(command);
    } catch (err) {
      throw new E_LOCK_NOT_OWNED();
    }
  }
};
export {
  DynamoDBStore,
  dynamodbStore
};
