// @verrou/core@0.5.1/build/src/drivers/knex.js
import {
  DatabaseStore
} from "../../chunk-MDY27ECV.js";
import "../../chunk-QYHBKL3W.js";

// src/drivers/knex.ts
function knexStore(config) {
  return {
    config,
    factory: () => {
      const adapter = new KnexAdapter(config.connection);
      return new DatabaseStore(adapter, config);
    }
  };
}
var KnexAdapter = class {
  #connection;
  #tableName;
  constructor(connection) {
    this.#connection = connection;
  }
  setTableName(tableName) {
    this.#tableName = tableName;
  }
  async createTableIfNotExists() {
    const hasTable = await this.#connection.schema.hasTable(this.#tableName);
    if (hasTable) return;
    await this.#connection.schema.createTable(this.#tableName, (table) => {
      table.string("key", 255).notNullable().primary();
      table.string("owner").notNullable();
      table.bigint("expiration").unsigned().nullable();
    });
  }
  async insertLock(lock) {
    await this.#connection.table(this.#tableName).insert(lock);
  }
  async acquireLock(lock) {
    const updated = await this.#connection.table(this.#tableName).where("key", lock.key).where("expiration", "<=", Date.now()).update({ owner: lock.owner, expiration: lock.expiration });
    return updated;
  }
  async deleteLock(key, owner) {
    await this.#connection.table(this.#tableName).where("key", key).modify((query) => {
      if (owner) query.where("owner", owner);
    }).delete();
  }
  async extendLock(key, owner, duration) {
    const updated = await this.#connection.table(this.#tableName).where("key", key).where("owner", owner).update({ expiration: Date.now() + duration });
    return updated;
  }
  async getLock(key) {
    const result = await this.#connection.table(this.#tableName).where("key", key).select(["owner", "expiration"]).first();
    return result;
  }
};
export {
  KnexAdapter,
  knexStore
};
