// @verrou/core@0.5.1/build/src/drivers/kysely.js
import {
  DatabaseStore
} from "../../chunk-MDY27ECV.js";
import "../../chunk-QYHBKL3W.js";

// src/drivers/kysely.ts
import { PostgresAdapter } from "kysely";
function kyselyStore(config) {
  return {
    config,
    factory: () => {
      const adapter = new KyselyAdapter(config.connection);
      return new DatabaseStore(adapter, config);
    }
  };
}
var KyselyAdapter = class {
  #connection;
  #tableName;
  constructor(connection) {
    this.#connection = connection;
  }
  setTableName(tableName) {
    this.#tableName = tableName;
  }
  async createTableIfNotExists() {
    const isPg = this.#connection.getExecutor().adapter instanceof PostgresAdapter;
    await this.#connection.schema.createTable(this.#tableName).addColumn("key", "varchar(255)", (col) => col.primaryKey().notNull()).addColumn("owner", "varchar(255)", (col) => col.notNull()).addColumn("expiration", "bigint", (col) => {
      if (!isPg) col.unsigned();
      return col;
    }).ifNotExists().execute();
  }
  async insertLock(lock) {
    await this.#connection.insertInto(this.#tableName).values({
      key: lock.key,
      owner: lock.owner,
      expiration: lock.expiration
    }).execute();
  }
  async acquireLock(lock) {
    const updated = await this.#connection.updateTable(this.#tableName).where("key", "=", lock.key).where("expiration", "<=", Date.now()).set({ owner: lock.owner, expiration: lock.expiration }).executeTakeFirst();
    return Number(updated.numUpdatedRows);
  }
  async deleteLock(key, owner) {
    await this.#connection.deleteFrom(this.#tableName).where("key", "=", key).$if(owner !== void 0, (query) => query.where("owner", "=", owner)).execute();
  }
  async extendLock(key, owner, duration) {
    const updated = await this.#connection.updateTable(this.#tableName).where("key", "=", key).where("owner", "=", owner).set({ expiration: Date.now() + duration }).executeTakeFirst();
    return Number(updated.numUpdatedRows);
  }
  async getLock(key) {
    const result = await this.#connection.selectFrom(this.#tableName).where("key", "=", key).select(["owner", "expiration"]).executeTakeFirst();
    return result;
  }
};
export {
  KyselyAdapter,
  kyselyStore
};
