// @verrou/core@0.5.1/build/src/drivers/memory.js
import {
  E_LOCK_NOT_OWNED
} from "../../chunk-QYHBKL3W.js";

// src/drivers/memory.ts
import { Mutex, tryAcquire } from "async-mutex";
function memoryStore() {
  return { factory: () => new MemoryStore() };
}
var MemoryStore = class {
  #locks = /* @__PURE__ */ new Map();
  /**
   * For a given key, get or create a new lock
   */
  getOrCreateForKey(key, owner) {
    let lock = this.#locks.get(key);
    if (!lock) {
      lock = { mutex: new Mutex(), owner };
      this.#locks.set(key, lock);
    }
    return lock;
  }
  /**
   * Compute the expiration date of a lock
   */
  #computeExpiresAt(ttl) {
    return ttl ? Date.now() + ttl : Number.POSITIVE_INFINITY;
  }
  /**
   * Check if lock is expired
   */
  #isLockEntryExpired(lock) {
    return lock.expiresAt && lock.expiresAt < Date.now();
  }
  /**
   * Extend the lock expiration. Throws an error if the lock is not owned by the owner
   * Duration is in milliseconds
   */
  async extend(key, owner, duration) {
    const lock = this.#locks.get(key);
    if (!lock || lock.owner !== owner) throw new E_LOCK_NOT_OWNED();
    lock.expiresAt = this.#computeExpiresAt(duration);
  }
  /**
   * Save the lock in the store if not already locked by another owner
   */
  async save(key, owner, ttl) {
    try {
      const lock = this.getOrCreateForKey(key, owner);
      if (this.#isLockEntryExpired(lock)) lock.releaser?.();
      lock.releaser = await tryAcquire(lock.mutex).acquire();
      lock.expiresAt = this.#computeExpiresAt(ttl);
      return true;
    } catch {
      return false;
    }
  }
  /**
   * Delete the lock from the store if it is owned by the owner
   * Otherwise throws a E_LOCK_NOT_OWNED error
   */
  async delete(key, owner) {
    const mutex = this.#locks.get(key);
    if (!mutex || !mutex.releaser) return;
    if (mutex.owner !== owner) throw new E_LOCK_NOT_OWNED();
    mutex.releaser();
    this.#locks.delete(key);
  }
  /**
   * Force delete the lock from the store. No check is made on the owner
   */
  async forceDelete(key) {
    const lock = this.#locks.get(key);
    if (!lock) return;
    lock.releaser?.();
  }
  /**
   * Check if the lock exists
   */
  async exists(key) {
    const lock = this.#locks.get(key);
    if (!lock || this.#isLockEntryExpired(lock)) return false;
    return lock.mutex.isLocked();
  }
};
export {
  MemoryStore,
  memoryStore
};
