// @verrou/core@0.5.1/build/src/drivers/redis.js
import {
  E_LOCK_NOT_OWNED
} from "../../chunk-QYHBKL3W.js";

// src/drivers/redis.ts
function redisStore(options) {
  return { factory: () => new RedisStore(options) };
}
var RedisStore = class {
  /**
   * IORedis connection instance
   */
  #connection;
  constructor(options) {
    this.#connection = options.connection;
  }
  /**
   * Save the lock in the store if not already locked by another owner
   */
  async save(key, owner, ttl) {
    if (ttl) {
      const result2 = await this.#connection.set(key, owner, "PX", ttl, "NX");
      return result2 === "OK";
    }
    const result = await this.#connection.setnx(key, owner);
    return result === 1;
  }
  /**
   * Delete the lock from the store if it is owned by the owner
   * Otherwise throws a E_LOCK_NOT_OWNED error
   */
  async delete(key, owner) {
    const lua = `
      if redis.call("get", KEYS[1]) == ARGV[1] then
        return redis.call("del", KEYS[1])
      else
        return 0
      end
    `;
    const result = await this.#connection.eval(lua, 1, key, owner);
    if (result === 0) throw new E_LOCK_NOT_OWNED();
  }
  /**
   * Force delete the lock from the store. No check is made on the owner
   */
  async forceDelete(key) {
    await this.#connection.del(key);
  }
  /**
   * Check if the lock exists
   */
  async exists(key) {
    const result = await this.#connection.get(key);
    return !!result;
  }
  /**
   * Extend the lock expiration. Throws an error if the lock is not owned by the owner
   * Duration is in milliseconds
   */
  async extend(key, owner, duration) {
    const lua = `
      if redis.call("get", KEYS[1]) == ARGV[1] then
        return redis.call("pexpire", KEYS[1], ARGV[2])
      else
        return 0
      end
    `;
    const result = await this.#connection.eval(lua, 1, key, owner, duration);
    if (result === 0) throw new E_LOCK_NOT_OWNED();
  }
};
export {
  RedisStore,
  redisStore
};
