// @poppinss/defer@1.1.2/build/index.js
// src/queue.ts
import fastq from "fastq";
var Deferred = class {
  resolve;
  reject;
  promise = new Promise((resolve, reject) => {
    this.reject = reject;
    this.resolve = resolve;
  });
};
var DeferQueue = class {
  /**
   * Custom drain callback assigned to the queue. We keep a
   * reference to it so that we can assign it back after
   * the kill method
   */
  #drainCallback;
  /**
   * Reference to the undelying fastq
   */
  #fastQueue;
  /**
   * Notifiers is a collection of promises that are invoked
   * when a task is completed and then removed from the
   * stack.
   *
   * You can create notifiers using the "createNotifier" method
   * to lineraly monitor the completion of a task
   */
  #notifiers = [];
  /**
   * Notifies the onError and taskCompleted handlers
   */
  #notifier = (error, taskResult) => {
    const awaitingNotifier = this.#notifiers.shift();
    if (error) {
      if (this.onError) {
        this.onError(error, taskResult);
      }
      if (awaitingNotifier) {
        awaitingNotifier.resolve({ status: "rejected", reason: error });
      }
      return;
    }
    if (this.taskCompleted) {
      this.taskCompleted(taskResult);
    }
    if (awaitingNotifier) {
      awaitingNotifier.resolve({ status: "fulfilled", value: taskResult });
    }
  };
  /**
   * Function that listens for the errors on the queue
   */
  onError;
  /**
   * Function that listens successful task completions
   */
  taskCompleted;
  /**
   * Function that listens for new task additions
   */
  taskAdded;
  /**
   * Function that is invoked when the last task is executed
   * by the worker
   */
  get drained() {
    return this.#fastQueue.drain;
  }
  set drained(callback) {
    this.#drainCallback = callback;
    this.#fastQueue.drain = callback;
  }
  constructor(options) {
    this.#fastQueue = fastq(async (cb, done) => {
      try {
        const response = await (typeof cb === "function" ? cb() : cb.run());
        done(null, { name: cb.name, response });
      } catch (error) {
        done(error, { name: cb.name });
      }
    }, options?.concurrency ?? 10);
  }
  /**
   * Create a notifier in advance that will be resolved when
   * a queued task is completed.
   *
   * Notifiers must be used during testing to track the execution
   * of an async task.
   *
   * One notifier can only listen for a maximum of one task that will
   * be executed after the notifier has been created and after that
   * the notifier will be removed.
   */
  createNotifier() {
    const deferred = new Deferred();
    this.#notifiers.push(deferred);
    return deferred.promise;
  }
  /**
   * Push a new callback (aka task) to the end of the queue.
   */
  push(task) {
    if (this.taskAdded) {
      this.taskAdded(task);
    }
    this.#fastQueue.push(task, this.#notifier);
    return this;
  }
  /**
   * Push a new callback (aka task) to the start of the queue.
   */
  unshift(task) {
    if (this.taskAdded) {
      this.taskAdded(task);
    }
    this.#fastQueue.unshift(task, this.#notifier);
    return this;
  }
  /**
   * Pause queue
   */
  pause() {
    this.#fastQueue.pause();
  }
  /**
   * Resume queue
   */
  resume() {
    this.#fastQueue.resume();
  }
  /**
   * Check if queue is currently idle
   */
  isIdle() {
    return this.#fastQueue.idle();
  }
  /**
   * Get the size of tasks in the queue
   */
  size() {
    return this.#fastQueue.length();
  }
  /**
   * Kill the queue and reset the drain method to a
   * noop function.
   *
   * The drain method will be called before the queue gets
   * killed
   */
  kill() {
    this.#fastQueue.killAndDrain();
    if (this.#drainCallback) {
      this.#fastQueue.drain = this.#drainCallback;
    }
  }
};
export {
  DeferQueue
};
