// @jrmc/adonis-attachment@5.0.3/build/src/adapters/meta.js
/**
 * @jrmc/adonis-attachment
 *
 * @license MIT
 * @copyright Jeremy Chaufourier <jeremy@chaufourier.fr>
 */
import path from 'node:path';
import fs from 'node:fs/promises';
import { fileTypeFromBuffer, fileTypeFromFile } from 'file-type';
import mime from 'mime-types';
function getFileExtension(filename) {
    if (!filename) {
        return '';
    }
    const ext = path.extname(filename).slice(1);
    return ext && /^[a-zA-Z0-9]+$/.test(ext) ? ext : '';
}
function metaByFileName(filename) {
    return {
        ext: getFileExtension(filename),
        mime: mime.lookup(filename) || '',
    };
}
export async function metaFormBuffer(input) {
    const fileType = await fileTypeFromBuffer(input);
    return {
        extname: fileType?.ext || '',
        mimeType: fileType?.mime || '',
        size: input.length,
    };
}
export async function metaFormFile(input, filename) {
    let fileType;
    let size = 0;
    fileType = metaByFileName(filename);
    if (fileType.ext === '' || fileType.mime === '') {
        fileType = await fileTypeFromFile(input);
    }
    const stats = await fs.stat(input);
    size = stats.size;
    return {
        extname: fileType?.ext || '',
        mimeType: fileType?.mime || '',
        size,
    };
}
